### DISCUSSION TITLE
thoughts on automating releases

### DISCUSSION THREAD

**Discussion by RoboMWM** [GriefPrevention/GriefPrevention#1150] [General]
More of just notes tracking how I want to go about this.

Initially started playing around with release drafter action, but I don't really label PRs.
- Maybe I should

I could also use that command Jikoo had, and just auto-replace with github usernames.

Could also have something that automatically generates a release when I make a modification to changelog.md (file doesn't exist yet).
- I'd like edits to changelog.md to also be propagated to any existing releases (e.g. when fixing typos)
  - May not be that important tho, and modifying earlier entries kinda breaks the audit trail anyways.

Appveyor has some stuff for releases too, so not really necessary to use a github action. https://www.appveyor.com/docs/deployment/github/

But ideally as aformentioned, would like a release to be triggered when pushing a change to changelog.md or pom.xml's version field.

---

Then this should also trigger upload to bukkitdev https://authors.curseforge.com/knowledge-base/projects/529-api and to maven central.

---

Automatic release drafter really isn't important other than for auto-sorting PRs into fixes and features (and improvements and other lesser-used subcategories as needed). Tagging PR # is helpful tho, especially since releases does that automatically - a changelog.md would not do this.

**Comment by RoboMWM** [comment]
https://github.com/ffurrer2/extract-release-notes may be interesting, hmm

**Reply by RoboMWM** [reply]
Only potential issue I see with this is that I don't ordinarily generate tags on the cli - I usually have github releases do that for me. Hmm. Well, another way could be taking a published release and making a changelog.md. That'd involve pushing to code tho, hence why I'd rather do it the other way.

**Reply by RoboMWM** [reply]
Another thing is to just have release drafter (or etc.) generate a release markdown as PRs and stuff are pulled in, and then when I'm ready to make a release I just take the last build's release notes artifact to modify as needed in changelog.md.

I could also get fancy and do some maven trickery so it performs the tagging n stuff for me without me thinking. hmm......

**Reply by Jikoo** [reply]
Not a huge fan of extract-release-notes because it looks [keepachangelog](like https://keepachangelog.com/en/1.0.0/) is a practice for people who have given up on writing good commit messages. I'll give them that not all commits are relevant (see: the last 2 from me where I fixed things I broke) but manually maintaining a changelog is what we're trying to avoid here. A bit of curation is a lot less work than a full manual writeup.

**Reply by RoboMWM** [reply]
Yea... think people like Z750 do changelogs tho so idk a good practice.

Part of the reasoning behind using a changelog.md is that github releases isn't portable (I thought the tag it creates would also include the contents written in the release, but it doesn't). That's the primary reason.

Regardless, what I currently do is manual, so no matter if I dumped it in a GitHub Release or changelog.md, the contents are the exact same.

**Comment by RoboMWM** [comment]
One ideal:

Pre-notes:
- I generate a list of features, fixes, etc. from Jikoo's git log, OR
- I get a list of pre-markdown-formatted PR references from the artifacts of (some runner, i.e. github actions release drafter).

Post-notes, pre-publish:
- I make a new entry in changelog.md. Specifying the version in a second-level header at the top, 3rd level and bullet points, etc. below it.
- I run (some script) that automatically creates a new tag based on the second-level header and `git push`es it
  - Could be a maven thing
  - Could be powershell
  - Could be bash
- (something) notices the new changelog.md entry and git tag, and drafts a github release
  - Maybe this is also possible via maven? or powershell? Then can negate need to have an action/runner monitoring for tag push

**Comment by RoboMWM** [comment]
Maybe a good start is to just implement the maven release plugin in the first place and then go from there.

**Comment by Jikoo** [comment]
GH actions can do CLI stuff for you. I'm eventually planning to set up an action for OpenInv that's only triggered from the Actions window to trip a release.
I already (theoretically - it hasn't been tested) [have release uploading running](https://github.com/lishid/OpenInv/blob/master/.github/workflows/ci.yml#L61), but it still requires me to manually tag the branch, which involves me manually editing values in a local script which I then set off manually. I also still fully manually write changelogs.

My script basically looks like this (from memory, not at the PC I'm afraid)

```bash
#!/bin/bash

version="W.X.Y"
snapshot="W.X.Z-SNAPSHOT"

mvn versions:set -DnewVersion=$version

git add .
git commit -m "Bump version to $version for release."
git tag -a $version

mvn clean package -am -P whatever

mvn versions:set -DnewVersion=$snapshot

git add .
git commit -m "Bump version to $snapshot for development."

git push --tags origin master
```

**Reply by Jikoo** [reply]
Oh and for reference comment on potential changelog generation from commits is [this post](https://github.com/TechFortress/GriefPrevention/pull/1122#issuecomment-735451911).

**Reply by Jikoo** [reply]
Relevant links for constructing GH names from users' email addresses:
https://stackoverflow.com/a/34600197
https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#commits
https://docs.github.com/en/free-pro-team@latest/rest/reference/search#constructing-a-search-query
(note to self: "q=\<email\> in:email")

Basically, I (eventually) plan to write a script that uses the git log to assemble a basic changelog. Then I'll pull the emails of the committers. If I can get a username for an email, I'll use `@<username>` otherwise they get a raw text mention via whatever their name was set to in git.

**Reply by Jikoo** [reply]
Re: constructing changelog automatically - I've finally gotten off my rear and written a script that does a basic automatic changelog from git history. It's introduced in lishid/OpenInv@1d5a836fd0b558c02b3f0fc44907e2ce3417f0c7 and (ideally - it's all untested so far) generates a draft release with commit history body when a new tag is pushed.

Combining that with a system based off of https://github.com/TechFortress/GriefPrevention/discussions/1217#discussioncomment-491274 I believe I'll be as close as possible to full automation. It would be a lot easier to not manually manage game versions the way the linked upload script does (game version could "easily" be [pulled from Maven](https://github.com/lishid/OpenInv/blob/1d5a836fd0b558c02b3f0fc44907e2ce3417f0c7/scripts/install_spigot_dependencies.sh#L28-L49), the action appears to allow using blob name) but it's definitely a great example of basic usage.

The only thing left would be automatic Spigot uploads, but I honestly never released OpenInv there because I hate their resource system. As a user I can't `wget` files because of the way they set up CloudFlare and it's harder to find things I'm interested in; as an author CurseForge seems to have far better discoverability. Due to the lack of mandatory moderation on Spigot there are just so many uploads of low quality garbage so fast that you never can get any time on the first page of the list unless you become part of the problem and do daily junky uploads.

**Comment by Jikoo** [comment]
Plopping this here because I know we've discussed GH releases vs tags in the past and I can't seem to find it.

Something I only just remembered because I've been using GitHub for git for so long: you can actually create more detailed tags. Tags can actually have your changelog in their message, you just need to manually make them from CLI or some other tool.

Doesn't prevent the need to manually curate the release notes, but it does mean you would only need to write them once. That said, writing them on GH is probably easier due to the shiny UI, and an action can then push them elsewhere from there either way.

**Reply by RoboMWM** [reply]
Yea, just not as easy to change tag messages after the fact and I don't think github does any markdown formatting for those.

**Reply by Jikoo** [reply]
Very true on both counts, yeah.

**Comment by RoboMWM** [comment]
Well since this discussion Github has a "generate release notes" button and it primarily covers the PRs, which is mostly how development is being performed on this repo. I wonder if there is a way to customize how that button works.

Other things to take care of are initiating releases on dev end, is `mvn deploy` a recommended route? And also to figure out how to use maven central...

**Reply by RoboMWM** [reply]
created new discussions for these:
- Build and release process: https://github.com/GriefPrevention/GriefPrevention/discussions/2493
- Maven central publishing: TBA

**Comment by RoboMWM** [comment]
Going to close this as I'm seeing less of a need to automate creating release notes:
- GitHub's "Generate Release Notes" button exists
- Intended slower release cycle
- I mentioned this somewhere, not sure where; I'm focusing my minecraft development efforts primarily on GriefPrevention; overhead of release notes is likely less than overhead of maintaining an automated release notes generator.
- AI exists and can potentially do/assist this in the future anyway? lol. 
- Provides opportunity for a more "human" interaction in writing an article for people to read about new stuff with GriefPrevention development.
